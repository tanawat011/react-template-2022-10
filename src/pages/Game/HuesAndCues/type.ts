/* istanbul ignore file */
import type { DocumentData, QueryDocumentSnapshot } from 'firebase/firestore'

export type Player = {
  id: string
  name: string
}

export type Room = {
  id: string
  password: string
  hintChoice: string[]
  hintSelected: string
  isStarted: boolean
  isSubmitResult: boolean
  totalRound: number
}

export type RawRoomPlayer = {
  refPlayer: QueryDocumentSnapshot<DocumentData>
  color: string
  score: number
  allSelected: string[]
  isOwner: boolean
  isTurn: boolean
  isHinter: boolean
  totalTurn: number
  seq: number
}

export type RoomPlayer = Omit<RawRoomPlayer, 'refPlayer'> & {
  player: Player
}
