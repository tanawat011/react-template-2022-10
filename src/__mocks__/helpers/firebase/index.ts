import type * as Firebase from 'helpers/firebase'

const firebase = jest.createMockFromModule<typeof Firebase>('helpers/firebase')

module.exports = firebase
