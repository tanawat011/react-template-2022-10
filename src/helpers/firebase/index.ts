export * from './deleteCollection'
export * from './deleteDocument'
export * from './getAllDocuments'
export * from './getDocument'
export * from './getDocumentWithRef'
export * from './setDocument'
export * from './subscribeCollection'
export * from './subscribeDocument'
export * from './updateDocument'
