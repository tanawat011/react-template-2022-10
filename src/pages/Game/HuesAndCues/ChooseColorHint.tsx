import type { RoomPlayer } from './_type'

import { Button } from 'components/Button'
import { TwRow } from 'components/Common'
import { huesAndCues } from 'data/HuesAndCues'

import { CellClickable } from './Common'

type Prop = {
  currentPlayer: RoomPlayer
  updateCurrentPlayer: (currentPlayer: RoomPlayer) => void
}

export const ChooseColorHint: React.FC<Prop> = ({ currentPlayer, updateCurrentPlayer }) => {
  const randomNumber = (toNumber = 10) => {
    return Math.floor(Math.random() * toNumber) + 1
  }

  const randomColorId = () => {
    const charNumber = randomNumber(16)
    const colNumber = randomNumber(30)
    const char = String.fromCharCode(charNumber + 64)

    return `${char}${colNumber}`
  }

  const handleRandomColorHint = async () => {
    if (currentPlayer) {
      let choice: string[] = []

      for (let i = 1; i <= 4; i++) {
        choice = [...choice, randomColorId()]
      }

      const _currentPlayer: RoomPlayer = {
        ...currentPlayer,
        choice,
      }

      updateCurrentPlayer(_currentPlayer)
    }
  }

  const handleChooseColorHint = async (result: string) => {
    if (currentPlayer && !currentPlayer?.result) {
      const _currentPlayer: RoomPlayer = {
        ...currentPlayer,
        result,
      }

      updateCurrentPlayer(_currentPlayer)
    }
  }

  const colorsSelectable = currentPlayer?.choice || []
  const colorSelected = huesAndCues.flat().find((c) => c.id === currentPlayer?.result)
  const hasColorSelectable = !!colorsSelectable.length
  const nonColorSelectable = !hasColorSelectable

  return (
    <div className='mb-2 p-3 shadow-around flex divide-x'>
      <div>
        <div className='pl-2 mb-1 text-white font-bold select-none'>Choose Color Hint</div>

        <TwRow>
          {nonColorSelectable && <Button onClick={handleRandomColorHint}>Random Color Hint</Button>}

          {hasColorSelectable &&
            colorsSelectable.map((colorId) => {
              const data = huesAndCues.flat().find((c) => c.id === colorId)

              return (
                <CellClickable
                  key={`choose-color-hint-${data?.color}`}
                  className={data?.color}
                  onClick={() => handleChooseColorHint(colorId)}
                >
                  {colorId}
                </CellClickable>
              )
            })}
        </TwRow>
      </div>

      <div className='ml-3 pl-3'>
        <div className='mb-1 text-white font-bold select-none'>Color Selected</div>

        <TwRow>
          {colorSelected && (
            <CellClickable
              key={`color-selected-${colorSelected?.color}`}
              className={colorSelected?.color}
            >
              {colorSelected?.id}
            </CellClickable>
          )}
        </TwRow>
      </div>
    </div>
  )
}
