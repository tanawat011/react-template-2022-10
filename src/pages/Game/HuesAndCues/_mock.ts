import type { RoomPlayer } from './_type'

const PLAYER_BASE: RoomPlayer = {
  cells: [],
  uid: '1',
  name: 'Player 1',
  score: 0,
  color: '',
  number: 1,
}

const HINTER: RoomPlayer = {
  uid: 'p1',
  name: 'p1',
  number: 1,
  color: 'bg-[#0758a5]',
  score: 0,
  hinter: true,
  cells: [],
}

const HINTER_HAS_CHOICE: RoomPlayer = {
  ...HINTER,
  choice: ['A1', 'A2', 'A3', 'A4'],
}

const HINTER_HAS_CHOICE_SELECT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE,
  result: 'A1',
}

const HINTER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT,
  isSubmitResult: true,
}

const HINTER_IS_OWNER: RoomPlayer = {
  ...HINTER,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE: RoomPlayer = {
  ...HINTER_HAS_CHOICE,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE_SELECT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT,
  owner: true,
}

const PLAYER: RoomPlayer = {
  uid: 'P2',
  name: 'p2',
  number: 2,
  color: 'bg-[#117822]',
  score: 0,
  cells: [],
}

const PLAYER_MY_TURN: RoomPlayer = {
  ...PLAYER,
  isMyTurn: true,
}

const PLAYER_MY_TURN_SELECT_CELL: RoomPlayer = {
  ...PLAYER_MY_TURN,
  cells: ['A1'],
}

const result = 'F4'
const choice = ['F4', 'A9', 'P27', 'E15']

export const MOCK: {
  ONE_PLAYER: RoomPlayer[]
  TWO_PLAYER: RoomPlayer[]
  TWO_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER: RoomPlayer[]
  THREE_PLAYER_SECOND_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_TURN: RoomPlayer[]
  HINTER_NO_CHOICE: RoomPlayer[]
  HINTER_NO_SELECT_CHOICE: RoomPlayer[]
  HINTER_SELECTED_CHOICE: RoomPlayer[]
  PLAYER_NO_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: RoomPlayer[]
} = {
  ONE_PLAYER: [
    {
      ...PLAYER_BASE,
      cells: ['D17'],
      number: 1,
      hinter: true,
      isMyTurn: true,
      color: 'bg-[#0758a5]',
    },
  ],
  TWO_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
  ],
  TWO_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
  ],
  THREE_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
      owner: true,
      score: 1,
    },
    {
      ...PLAYER_BASE,
      uid: '2',
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      uid: '3',
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_TURN: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      isMyTurn: true,
    },
  ],
  THREE_PLAYER_SECOND_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      hinter: true,
    },
  ],
  HINTER_NO_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_NO_SELECT_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_SELECTED_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
      result,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_NO_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      color: 'bg-[#117822]',
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR_START_GAME: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      result: 'F4',
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      cells: ['F4'],
    },
    {
      ...PLAYER_BASE,
      number: 3,
      cells: ['F5', 'F6'],
    },
  ],
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      isSubmitResult: true,
      countRound: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
}
