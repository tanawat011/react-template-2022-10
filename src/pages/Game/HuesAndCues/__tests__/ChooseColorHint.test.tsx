import type { RenderResult } from '@testing-library/react'

import { act, fireEvent, screen, waitFor } from '@testing-library/react'

import { renderWithProviders } from 'helpers/test'

import { ChooseColorHint } from '../ChooseColorHint'
import { MOCK } from '../_mock'

jest.mock('../Common', () => ({
  ...jest.requireActual('../Common'),
  CellClickable: (props: React.HTMLAttributes<never>) => {
    return (
      <div {...props} data-testid={props.className}>
        {props.children}
      </div>
    )
  },
}))

describe('<ChooseColorHint />', () => {
  test('match snapshot', () => {
    const updateCurrentPlayer = jest.fn()

    const { asFragment } = renderWithProviders(
      <ChooseColorHint
        currentPlayer={MOCK.HINTER_NO_CHOICE[0]}
        updateCurrentPlayer={updateCurrentPlayer}
      />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('current player click random 4 color for hint', async () => {
    let el: RenderResult

    const updateCurrentPlayer = jest.fn()

    await act(async () => {
      el = renderWithProviders(
        <ChooseColorHint
          currentPlayer={MOCK.HINTER_NO_CHOICE[0]}
          updateCurrentPlayer={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('current player selecting choice', async () => {
    let el: RenderResult

    const updateCurrentPlayer = jest.fn()

    await act(async () => {
      el = renderWithProviders(
        <ChooseColorHint
          currentPlayer={MOCK.HINTER_NO_SELECT_CHOICE[0]}
          updateCurrentPlayer={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getAllByTestId('bg-[#fbab2b]')[0]
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('current player selecting choice, but has choice', async () => {
    let el: RenderResult

    const updateCurrentPlayer = jest.fn()

    await act(async () => {
      el = renderWithProviders(
        <ChooseColorHint
          currentPlayer={MOCK.HINTER_SELECTED_CHOICE[0]}
          updateCurrentPlayer={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getAllByTestId('bg-[#fbab2b]')[0]
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('current player selected choice', async () => {
    const updateCurrentPlayer = jest.fn()

    const { asFragment } = renderWithProviders(
      <ChooseColorHint
        currentPlayer={MOCK.HINTER_SELECTED_CHOICE[0]}
        updateCurrentPlayer={updateCurrentPlayer}
      />,
    )

    expect(asFragment()).toBeTruthy()
  })
})
