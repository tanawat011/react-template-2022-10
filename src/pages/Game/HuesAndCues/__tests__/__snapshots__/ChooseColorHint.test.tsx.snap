// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseColorHint /> match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<div
    class="mb-2 p-3 shadow-around flex divide-x"
  >
    <div>
      <div
        class="pl-2 mb-1 text-white font-bold select-none"
      >
        Choose Color Hint
      </div>
      <div
        class="c0"
      >
        <button
          data-testid="button"
        >
          Random Color Hint
        </button>
      </div>
    </div>
    <div
      class="ml-3 pl-3"
    >
      <div
        class="mb-1 text-white font-bold select-none"
      >
        Color Selected
      </div>
      <div
        class="c0"
      />
    </div>
  </div>
</DocumentFragment>
`;
